src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] values;
+
+    public ArrayList() {
+        this.values = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= values.length) {
+            values = grow(values.length);
+        }
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkBoundsIndex((index == size && size != 0) ? (index - 1) : index);
+        int newSize = size + 1;
+        if (newSize >= values.length) {
+            values = grow(values.length);
+        }
+        T[] tempArray = (T[]) new Object[values.length];
+        System.arraycopy(values,0, tempArray, 0, index);
+        tempArray[index] = value;
+        System.arraycopy(values, index, tempArray, index + 1, size - index);
+        values = tempArray;
+        size = newSize;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
-    public T get(int index) {
+    public T get(int index) throws ArrayListIndexOutOfBoundsException {
+        checkBoundsIndex(index);
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                return ((values[i] == null) ? null : (T) values[i]);
+            }
+        }
         return null;
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public void set(T value, int index) throws ArrayListIndexOutOfBoundsException {
+        checkBoundsIndex(index);
+        for (int i = 0; i < size; i++) {
+            if (i == index) {
+                values[i] = value;
+            }
+        }
     }
 
     @Override
-    public T remove(int index) {
-        return null;
+    public T remove(int index) throws NoSuchElementException {
+        checkBoundsIndex(index);
+        T[] tempValues = (T[]) new Object[values.length];
+        System.arraycopy(values, 0, tempValues, 0, index);
+        System.arraycopy(values, index + 1, tempValues, index, size - index - 1);
+        T oldValue = get(index);
+        values = tempValues;
+        size--;
+        return oldValue;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) throws NoSuchElementException {
+        T foundValue = null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if ((element == null && get(i) == null)
+                    || ((element != null && get(i) != null) && get(i).equals(element))) {
```

Comment:
```suggestion
            if (element == values[i] || element != null && element.equals(values[i])) {
```
we can do it simpler, as in Objects.equals impl